IdentityCore/src/util/MSIDKeyOperationUtil.m

```diff
@@ -0,0 +1,129 @@
+//
+//  MSIDKeyOperationUtil.m
+//  IdentityCore
+//
+//  Created by Ameya Patil on 9/24/21.
+//  Copyright © 2021 Microsoft. All rights reserved.
+//
+
+#import "MSIDKeyOperationUtil.h"
+#import "MSIDJwtAlgorithm.h"
+#import "NSData+JWT.h"
+#import "NSData+MSIDExtensions.h"
+
+@implementation MSIDKeyOperationUtil
+
++ (MSIDKeyOperationUtil *)sharedInstance
+{
+    static dispatch_once_t once;
+    static MSIDKeyOperationUtil *s_keyOperationUtil = nil;
+    dispatch_once(&once, ^{
+        s_keyOperationUtil = [[MSIDKeyOperationUtil alloc] init];
+    });
+    return s_keyOperationUtil;
+}
+
+- (BOOL)isKeyFromSecureEnclave:(SecKeyRef)key
+{
+    if (key)
+    {
+        NSDictionary *attributes = (NSDictionary *)CFBridgingRelease(SecKeyCopyAttributes(key));
+        NSString *attrTokenId = (NSString *)[attributes valueForKey:(NSString *)CFBridgingRelease(kSecAttrTokenID)];
+        NSString *secureEnclaveId = (NSString *)(CFBridgingRelease(kSecAttrTokenIDSecureEnclave));  // id = com.apple.settoken for key in secure enclave
+        return [secureEnclaveId isEqualToString:attrTokenId];
+    }
+    return NO;
+}
+
+- (BOOL)isOperationSupportedByKey:(SecKeyOperationType)operation algorithm:(SecKeyAlgorithm)algorithm key:(SecKeyRef)key context:(id<MSIDRequestContext> _Nullable)context error:(NSError * _Nullable __autoreleasing *)error
+{
+    if (key == NULL)
+    {
+        [self generateErrorWithMessage:@"Key passed in to check operation is not defined." underlyingError:nil context:context error:error];
+        return NO;
+    }
+    BOOL isSupported = SecKeyIsAlgorithmSupported(key, operation, algorithm);
+    if (!isSupported)
```

Comment:
nit: can we combine two if statements here. if (!key || !algorithm || !SecKeyIsAlgorithmSupported(key, operation, algorithm)?

Reply:
I want to define a different error message for different conditions so that it is easier to diagnose, hence kept it separate.